File: scope104/src/main/java/com/petro/scope104/presentation/list/WorkerListFragment.java
package com.petro.scope104.presentation.list;

import android.app.ActivityOptions;
import android.content.res.Configuration;
import android.os.Bundle;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.petro.scope104.R;
import com.petro.scope104.databinding.FragmentWorkersBinding;
import com.petro.scope104.domain.entity.WorkerEntity;
import com.petro.scope104.util.PaginationScrollListener;

import java.util.ArrayList;
import java.util.Set;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class WorkerListFragment extends Fragment {
    private static final String KEY_TYPE = "KEY_TYPE";
    private static final String KEY_DATA = "KEY_DATA";
    private final boolean isLastPage = false;
    private WorkerListAdapter adapter;
    private FragmentWorkersBinding binding;
    private WorkerListViewModel viewModel;

    public static Fragment newInstance(ListType type) {
        Bundle args = new Bundle();
        args.putSerializable(KEY_TYPE, type);
        WorkerListFragment fragment = new WorkerListFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putSerializable(KEY_DATA, new ArrayList<>(adapter.getCurrentList()));
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        viewModel = new ViewModelProvider(this).get(WorkerListViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = FragmentWorkersBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        assert getArguments() != null;
        ListType listType = ((ListType) getArguments().getSerializable(KEY_TYPE));
        RecyclerView rv = binding.listOfWorkers;
        final SwipeRefreshLayout pullToRefresh = binding.pullToRefresh;
        pullToRefresh.setColorSchemeColors(ContextCompat.getColor(requireContext(), R.color.purple_500));
        pullToRefresh.setOnRefreshListener(() -> {
            loadMore(true);
            pullToRefresh.setRefreshing(false);
        });
        int spanCount = 3;
        if (getResources().getConfiguration().orientation == Configuration.ORIENTATION_LANDSCAPE) {
            spanCount = 6;
        }
        switch (listType) {
            case LINEAR:
                adapter = new WorkerListAdapter(WorkerListAdapter.VIEW_TYPE_NORMAL);
                rv.setLayoutManager(new LinearLayoutManager(getContext()));
                break;
            case GRID:
                adapter = new WorkerListAdapter(WorkerListAdapter.VIEW_TYPE_SQUARE);
                rv.setLayoutManager(new GridLayoutManager(getContext(), spanCount));
                break;
            case STAGGERED:
                adapter = new WorkerListAdapter(WorkerListAdapter.VIEW_TYPE_SQUARE);
                rv.setLayoutManager(new StaggeredGridLayoutManager(spanCount, StaggeredGridLayoutManager.VERTICAL));
                break;
        }
        rv.setAdapter(adapter);
        adapter.setOnClickListener((clickedItem, transitionView) -> {
            Pair<View, String> p1 = Pair.create(transitionView.get(0), transitionView.get(0).getTransitionName());
            Pair<View, String> p2 = Pair.create(transitionView.get(1), transitionView.get(1).getTransitionName());
            ActivityOptions activityOptions = ActivityOptions.makeSceneTransitionAnimation(getActivity(), p1, p2);
            ((WorkerListInteractions) requireActivity()).onItemClick(clickedItem, activityOptions);
        });
        if (savedInstanceState == null) {
            loadMore(false);
        }

        if (rv.getLayoutManager() instanceof LinearLayoutManager) {
            LinearLayoutManager linearLayoutManager = ((LinearLayoutManager) rv.getLayoutManager());
            rv.addOnScrollListener(new PaginationScrollListener(linearLayoutManager) {
                @Override
                protected void loadMoreItems() {
                    loadMore(false);
                }

                @Override
                public boolean isLastPage() {
                    return isLastPage;
                }

                @Override
                public boolean isLoading() {
                    return Boolean.TRUE.equals(viewModel.getIsLoadingLiveData().getValue());
                }
            });
        }

        viewModel.getIsLoadingLiveData().observe(getViewLifecycleOwner(), this::setLoading);
        viewModel.getUsers().observe(getViewLifecycleOwner(), adapter::submitList);
    }

    public void refresh() {
        loadMore(true);
    }

    private void loadMore(boolean refresh) {
        if (getActivity() instanceof ListFilter) {
            ListFilter listFilter = (ListFilter) getActivity();
            viewModel.loadMore(refresh, listFilter.getCurrentSelectedGender(), listFilter.getCountries());
        }
    }

    private void setLoading(boolean isLoading) {
        binding.progress.setVisibility(isLoading ? View.VISIBLE : View.GONE);
    }

    public interface ListFilter {
        Gender getCurrentSelectedGender();

        Set<String> getCountries();
    }

    interface WorkerListInteractions {
        void onItemClick(WorkerEntity workerEntity, ActivityOptions activityOptions);
    }
}

File: scope104/src/main/kotlin/com/petro/scope104/presentation/list/WorkerListFragment.kt
package com.petro.scope104.presentation.list

import android.app.ActivityOptions
import android.content.res.Configuration
import android.os.Bundle
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.petro.scope104.R
import com.petro.scope104.databinding.FragmentWorkersBinding
import com.petro.scope104.domain.entity.WorkerEntity
import com.petro.scope104.util.PaginationScrollListener
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class WorkerListFragment : Fragment() {
    private val isLastPage = false
    private lateinit var adapter: WorkerListAdapter
    private lateinit var binding: FragmentWorkersBinding
    private lateinit var viewModel: WorkerListViewModel
    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putSerializable(
            KEY_DATA, ArrayList(
                adapter.currentList
            )
        )
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        viewModel = ViewModelProvider(this)[WorkerListViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = FragmentWorkersBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        val listType = requireArguments().getSerializable(KEY_TYPE) as ListType
        val rv = binding.listOfWorkers
        val pullToRefresh = binding.pullToRefresh
        pullToRefresh.setColorSchemeColors(
            ContextCompat.getColor(
                requireContext(),
                R.color.purple_500
            )
        )
        pullToRefresh.setOnRefreshListener {
            loadMore(true)
            pullToRefresh.isRefreshing = false
        }
        var spanCount = 3
        if (resources.configuration.orientation == Configuration.ORIENTATION_LANDSCAPE) {
            spanCount = 6
        }
        when (listType) {
            ListType.LINEAR -> {
                adapter = WorkerListAdapter(WorkerListAdapter.VIEW_TYPE_NORMAL)
                rv.layoutManager = LinearLayoutManager(context)
            }
            ListType.GRID -> {
                adapter = WorkerListAdapter(WorkerListAdapter.VIEW_TYPE_SQUARE)
                rv.layoutManager = GridLayoutManager(context, spanCount)
            }
            ListType.STAGGERED -> {
                adapter = WorkerListAdapter(WorkerListAdapter.VIEW_TYPE_SQUARE)
                rv.layoutManager =
                    StaggeredGridLayoutManager(spanCount, StaggeredGridLayoutManager.VERTICAL)
            }
        }
        rv.adapter = adapter
        adapter.setOnClickListener(object : WorkerListAdapter.OnClickListener{
            override fun onClick(clickedItem: WorkerEntity, transitionView: List<View>) {
                val p1 = Pair.create(
                    transitionView[0], transitionView[0].transitionName
                )
                val p2 = Pair.create(transitionView[1], transitionView[1].transitionName)
                val activityOptions = ActivityOptions.makeSceneTransitionAnimation(activity, p1, p2)
                (requireActivity() as WorkerListInteractions).onItemClick(clickedItem, activityOptions)            }
        } )

        if (savedInstanceState == null) {
            loadMore(false)
        }
        if (rv.layoutManager is LinearLayoutManager) {
            val linearLayoutManager = rv.layoutManager as LinearLayoutManager?
            rv.addOnScrollListener(object : PaginationScrollListener(linearLayoutManager) {
                override fun loadMoreItems() {
                    loadMore(false)
                }

                override val isLastPage: Boolean get() = this@WorkerListFragment.isLastPage
                override val isLoading: Boolean
                    get() = java.lang.Boolean.TRUE == viewModel.isLoadingLiveData.value
            })
        }
        viewModel.isLoadingLiveData
            .observe(viewLifecycleOwner) { isLoading: Boolean -> setLoading(isLoading) }
        viewModel.users
            .observe(viewLifecycleOwner) { list: List<WorkerEntity?> -> adapter.submitList(list) }
    }

    fun refresh() {
        loadMore(true)
    }

    private fun loadMore(refresh: Boolean) {
        if (activity is ListFilter) {
            val listFilter = activity as ListFilter?
            viewModel.loadMore(refresh, listFilter!!.currentSelectedGender, listFilter.countries)
        }
    }

    private fun setLoading(isLoading: Boolean) {
        binding.progress.visibility = if (isLoading) View.VISIBLE else View.GONE
    }

    interface ListFilter {
        val currentSelectedGender: Gender
        val countries: Set<String>
    }

    internal interface WorkerListInteractions {
        fun onItemClick(workerEntity: WorkerEntity?, activityOptions: ActivityOptions)
    }

    companion object {
        private const val KEY_TYPE = "KEY_TYPE"
        private const val KEY_DATA = "KEY_DATA"
        fun newInstance(type: ListType?): Fragment {
            val args = Bundle()
            args.putSerializable(KEY_TYPE, type)
            val fragment = WorkerListFragment()
            fragment.arguments = args
            return fragment
        }
    }
}